# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 7e9d56eb485e
# Full Hash: 7e9d56eb485e06cca2a82fe88e2ae36c2ccd2095
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2020-04-01 21:26:59
# Regressor Bug: 1623278
# File Overlap Count: 1
# Description:
#   Bug 1623278 - Use SafeRefPtr for FileManager. r=dom-workers-and-storage-reviewers,asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D65549
# ==============================================================================

diff -r a753c018a4a8 -r 7e9d56eb485e dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Wed Apr 01 09:54:24 2020 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Wed Apr 01 09:52:54 2020 +0000
@@ -3414,7 +3414,7 @@
 }
 
 class UpgradeFileIdsFunction final : public mozIStorageFunction {
-  RefPtr<FileManager> mFileManager;
+  SafeRefPtr<FileManager> mFileManager;
 
  public:
   UpgradeFileIdsFunction() { AssertIsOnIOThread(); }
@@ -4720,7 +4720,7 @@
  private:
   InitializedOnceMustBeTrue<const nsCOMPtr<mozIStorageConnection>>
       mStorageConnection;
-  InitializedOnceMustBeTrue<const RefPtr<FileManager>> mFileManager;
+  InitializedOnceMustBeTrue<const SafeRefPtr<FileManager>> mFileManager;
   nsInterfaceHashtable<nsCStringHashKey, mozIStorageStatement>
       mCachedStatements;
   RefPtr<UpdateRefcountFunction> mUpdateRefcountFunction;
@@ -4795,8 +4795,8 @@
   void EnableQuotaChecks();
 
  private:
-  DatabaseConnection(mozIStorageConnection* aStorageConnection,
-                     FileManager* aFileManager);
+  DatabaseConnection(nsCOMPtr<mozIStorageConnection> aStorageConnection,
+                     SafeRefPtr<FileManager> aFileManager);
 
   ~DatabaseConnection();
 
@@ -4881,7 +4881,7 @@
   enum class UpdateType { Increment, Decrement };
 
   DatabaseConnection* const mConnection;
-  FileManager* const mFileManager;
+  FileManager& mFileManager;
   nsClassHashtable<nsUint64HashKey, FileInfoEntry> mFileInfoEntries;
   nsDataHashtable<nsUint64HashKey, FileInfoEntry*> mSavepointEntriesIndex;
 
@@ -4896,7 +4896,7 @@
   NS_DECL_MOZISTORAGEFUNCTION
 
   UpdateRefcountFunction(DatabaseConnection* aConnection,
-                         FileManager* aFileManager);
+                         FileManager& aFileManager);
 
   nsresult WillCommit();
 
@@ -5892,7 +5892,7 @@
  private:
   RefPtr<Factory> mFactory;
   SafeRefPtr<FullDatabaseMetadata> mMetadata;
-  RefPtr<FileManager> mFileManager;
+  SafeRefPtr<FileManager> mFileManager;
   RefPtr<DirectoryLock> mDirectoryLock;
   nsTHashtable<nsPtrHashKey<TransactionBase>> mTransactions;
   nsTHashtable<nsPtrHashKey<MutableFile>> mMutableFiles;
@@ -5926,7 +5926,7 @@
            const Maybe<ContentParentId>& aOptionalContentParentId,
            const nsACString& aGroup, const nsACString& aOrigin,
            uint32_t aTelemetryId, SafeRefPtr<FullDatabaseMetadata> aMetadata,
-           RefPtr<FileManager> aFileManager,
+           SafeRefPtr<FileManager> aFileManager,
            RefPtr<DirectoryLock> aDirectoryLock, bool aFileHandleDisabled,
            bool aChromeWriteAccessAllowed);
 
@@ -5966,7 +5966,12 @@
 
   const nsString& FilePath() const { return mFilePath; }
 
-  FileManager* GetFileManager() const { return mFileManager; }
+  FileManager& GetFileManager() const { return *mFileManager; }
+
+  SafeRefPtr<FileManager> GetFileManagerPtr() const {
+    MOZ_ASSERT(mFileManager);
+    return mFileManager.clonePtr();
+  }
 
   const FullDatabaseMetadata& Metadata() const {
     MOZ_ASSERT(mMetadata);
@@ -6968,7 +6973,7 @@
   SafeRefPtr<FullDatabaseMetadata> mMetadata;
 
   uint64_t mRequestedVersion;
-  RefPtr<FileManager> mFileManager;
+  SafeRefPtr<FileManager> mFileManager;
 
   RefPtr<Database> mDatabase;
   RefPtr<VersionChangeTransaction> mVersionChangeTransaction;
@@ -7276,7 +7281,7 @@
 
   const IndexMetadata mMetadata;
   Maybe<UniqueIndexTable> mMaybeUniqueIndexTable;
-  const RefPtr<FileManager> mFileManager;
+  const SafeRefPtr<FileManager> mFileManager;
   const nsCString mDatabaseId;
   const IndexOrObjectStoreId mObjectStoreId;
 
@@ -8097,7 +8102,7 @@
  protected:
   explicit ValueCursorBase(TransactionBase* const aTransaction)
       : mDatabase(aTransaction->GetDatabase()),
-        mFileManager(mDatabase->GetFileManager()),
+        mFileManager(mDatabase->GetFileManagerPtr()),
         mBackgroundParent(aTransaction->GetBackgroundParent()) {
     MOZ_ASSERT(mDatabase);
     MOZ_ASSERT(mFileManager);
@@ -8108,7 +8113,7 @@
   ~ValueCursorBase() { MOZ_ASSERT(!mBackgroundParent); }
 
   const RefPtr<Database> mDatabase;
-  const RefPtr<FileManager> mFileManager;
+  const SafeRefPtr<FileManager> mFileManager;
 
   InitializedOnceMustBeTrue<PBackgroundParent* const> mBackgroundParent;
 };
@@ -8713,13 +8718,14 @@
   };
 
   nsCOMPtr<nsIEventTarget> mOwningEventTarget;
-  RefPtr<FileManager> mFileManager;
+  SafeRefPtr<FileManager> mFileManager;
   RefPtr<DirectoryLock> mDirectoryLock;
   nsTArray<int64_t> mFileIds;
   State mState;
 
  public:
-  DeleteFilesRunnable(FileManager* aFileManager, nsTArray<int64_t>&& aFileIds);
+  DeleteFilesRunnable(SafeRefPtr<FileManager> aFileManager,
+                      nsTArray<int64_t>&& aFileIds);
 
   void RunImmediately();
 
@@ -9103,7 +9109,7 @@
 // Then, FileManager::Get(Journal)Directory and FileManager::GetFileForId might
 // eventually be made private.
 class MOZ_STACK_CLASS FileHelper final {
-  const RefPtr<FileManager> mFileManager;
+  const SafeRefPtr<FileManager> mFileManager;
 
   InitializedOnceMustBeTrue<const nsCOMPtr<nsIFile>, LazyInit::Allow>
       mFileDirectory;
@@ -9115,7 +9121,7 @@
       mReadCallback;
 
  public:
-  explicit FileHelper(RefPtr<FileManager>&& aFileManager)
+  explicit FileHelper(SafeRefPtr<FileManager>&& aFileManager)
       : mFileManager(std::move(aFileManager)) {
     MOZ_ASSERT(mFileManager);
   }
@@ -9259,9 +9265,8 @@
     return result;
   }
 
-  FileManager* fileManager = aDatabase->GetFileManager();
-
-  nsCOMPtr<nsIFile> directory = fileManager->GetCheckedDirectory();
+  nsCOMPtr<nsIFile> directory =
+      aDatabase->GetFileManager().GetCheckedDirectory();
   if (NS_WARN_IF(!directory)) {
     IDB_REPORT_INTERNAL_ERR();
     return Err(NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR);
@@ -10244,7 +10249,7 @@
       IndexedDatabaseManager* const mgr = IndexedDatabaseManager::Get();
       MOZ_ASSERT(mgr);
 
-      const RefPtr<FileManager> fileManager =
+      const SafeRefPtr<FileManager> fileManager =
           mgr->GetFileManager(aPersistenceType, aOrigin, aDatabaseName);
 
       if (fileManager) {
@@ -10719,9 +10724,10 @@
  ******************************************************************************/
 
 DatabaseConnection::DatabaseConnection(
-    mozIStorageConnection* aStorageConnection, FileManager* aFileManager)
-    : mStorageConnection(aStorageConnection),
-      mFileManager(aFileManager),
+    nsCOMPtr<mozIStorageConnection> aStorageConnection,
+    SafeRefPtr<FileManager> aFileManager)
+    : mStorageConnection(std::move(aStorageConnection)),
+      mFileManager(std::move(aFileManager)),
       mInReadTransaction(false),
       mInWriteTransaction(false)
 #ifdef DEBUG
@@ -10730,8 +10736,8 @@
 #endif
 {
   AssertIsOnConnectionThread();
-  MOZ_ASSERT(aStorageConnection);
-  MOZ_ASSERT(aFileManager);
+  MOZ_ASSERT(mStorageConnection);
+  MOZ_ASSERT(mFileManager);
 }
 
 DatabaseConnection::~DatabaseConnection() {
@@ -10846,7 +10852,7 @@
     MOZ_ASSERT(mFileManager);
 
     RefPtr<UpdateRefcountFunction> function =
-        new UpdateRefcountFunction(this, mFileManager->get());
+        new UpdateRefcountFunction(this, **mFileManager);
 
     rv = (*mStorageConnection)
              ->CreateFunction(NS_LITERAL_CSTRING("update_refcount"),
@@ -11565,13 +11571,12 @@
 }
 
 DatabaseConnection::UpdateRefcountFunction::UpdateRefcountFunction(
-    DatabaseConnection* const aConnection, FileManager* const aFileManager)
+    DatabaseConnection* const aConnection, FileManager& aFileManager)
     : mConnection(aConnection),
       mFileManager(aFileManager),
       mInSavepoint(false) {
   MOZ_ASSERT(aConnection);
   aConnection->AssertIsOnConnectionThread();
-  MOZ_ASSERT(aFileManager);
 }
 
 nsresult DatabaseConnection::UpdateRefcountFunction::WillCommit() {
@@ -11722,7 +11727,7 @@
     return rv;
   }
 
-  auto filesOrErr = DeserializeStructuredCloneFiles(*mFileManager, ids);
+  auto filesOrErr = DeserializeStructuredCloneFiles(mFileManager, ids);
   if (NS_WARN_IF(filesOrErr.isErr())) {
     return filesOrErr.unwrapErr();
   }
@@ -11762,7 +11767,7 @@
   AUTO_PROFILER_LABEL(
       "DatabaseConnection::UpdateRefcountFunction::CreateJournals", DOM);
 
-  nsCOMPtr<nsIFile> journalDirectory = mFileManager->GetJournalDirectory();
+  nsCOMPtr<nsIFile> journalDirectory = mFileManager.GetJournalDirectory();
   if (NS_WARN_IF(!journalDirectory)) {
     return NS_ERROR_FAILURE;
   }
@@ -11792,7 +11797,7 @@
   AUTO_PROFILER_LABEL(
       "DatabaseConnection::UpdateRefcountFunction::RemoveJournals", DOM);
 
-  nsCOMPtr<nsIFile> journalDirectory = mFileManager->GetJournalDirectory();
+  nsCOMPtr<nsIFile> journalDirectory = mFileManager.GetJournalDirectory();
   if (NS_WARN_IF(!journalDirectory)) {
     return NS_ERROR_FAILURE;
   }
@@ -12107,8 +12112,8 @@
       return rv;
     }
 
-    connection =
-        new DatabaseConnection(storageConnection, aDatabase->GetFileManager());
+    connection = new DatabaseConnection(storageConnection,
+                                        aDatabase->GetFileManagerPtr());
 
     rv = connection->Init();
     if (NS_WARN_IF(NS_FAILED(rv))) {
@@ -13708,7 +13713,7 @@
                    const nsACString& aGroup, const nsACString& aOrigin,
                    uint32_t aTelemetryId,
                    SafeRefPtr<FullDatabaseMetadata> aMetadata,
-                   RefPtr<FileManager> aFileManager,
+                   SafeRefPtr<FileManager> aFileManager,
                    RefPtr<DirectoryLock> aDirectoryLock,
                    bool aFileHandleDisabled, bool aChromeWriteAccessAllowed)
     : mFactory(std::move(aFactory)),
@@ -16602,8 +16607,10 @@
     // 0, but the dbRefCnt is non-zero, which will keep the FileInfo object
     // alive.
     MOZ_ASSERT(dbRefCnt > 0);
-    mFileInfos.Put(id, new FileInfo(FileManagerGuard{}, this, id,
-                                    static_cast<nsrefcnt>(dbRefCnt)));
+    mFileInfos.Put(
+        id, new FileInfo(FileManagerGuard{},
+                         SafeRefPtr{this, AcquireStrongRefFromRawPtr{}}, id,
+                         static_cast<nsrefcnt>(dbRefCnt)));
 
     mLastFileId = std::max(id, mLastFileId);
   }
@@ -17672,8 +17679,8 @@
     const auto& value = pendingDeleteInfoEntry.GetData();
     MOZ_ASSERT(!value->IsEmpty());
 
-    RefPtr<DeleteFilesRunnable> runnable =
-        new DeleteFilesRunnable(key, std::move(*value));
+    RefPtr<DeleteFilesRunnable> runnable = new DeleteFilesRunnable(
+        SafeRefPtr{key, AcquireStrongRefFromRawPtr{}}, std::move(*value));
 
     MOZ_ASSERT(value->IsEmpty());
 
@@ -17919,11 +17926,11 @@
  * DeleteFilesRunnable
  ******************************************************************************/
 
-DeleteFilesRunnable::DeleteFilesRunnable(FileManager* aFileManager,
+DeleteFilesRunnable::DeleteFilesRunnable(SafeRefPtr<FileManager> aFileManager,
                                          nsTArray<int64_t>&& aFileIds)
     : Runnable("dom::indexeddb::DeleteFilesRunnable"),
       mOwningEventTarget(GetCurrentThreadEventTarget()),
-      mFileManager(aFileManager),
+      mFileManager(std::move(aFileManager)),
       mFileIds(std::move(aFileIds)),
       mState(State_Initial) {}
 
@@ -19306,8 +19313,8 @@
   // This file manager doesn't need real origin info, etc. The only purpose is
   // to store file ids without adding more complexity or code duplication.
   auto fileManager =
-      MakeRefPtr<FileManager>(PERSISTENCE_TYPE_INVALID, EmptyCString(),
-                              EmptyCString(), EmptyString(), false);
+      MakeSafeRefPtr<FileManager>(PERSISTENCE_TYPE_INVALID, EmptyCString(),
+                                  EmptyCString(), EmptyString(), false);
 
   nsresult rv = fileManager->Init(aFMDirectory, aConnection);
   if (NS_WARN_IF(NS_FAILED(rv))) {
@@ -21437,18 +21444,18 @@
   IndexedDatabaseManager* mgr = IndexedDatabaseManager::Get();
   MOZ_ASSERT(mgr);
 
-  RefPtr<FileManager> fileManager =
+  SafeRefPtr<FileManager> fileManager =
       mgr->GetFileManager(persistenceType, mOrigin, databaseName);
   if (!fileManager) {
-    fileManager = new FileManager(persistenceType, mGroup, mOrigin,
-                                  databaseName, mEnforcingQuota);
+    fileManager = MakeSafeRefPtr<FileManager>(persistenceType, mGroup, mOrigin,
+                                              databaseName, mEnforcingQuota);
 
     rv = fileManager->Init(fmDirectory, connection);
     if (NS_WARN_IF(NS_FAILED(rv))) {
       return rv;
     }
 
-    mgr->AddFileManager(fileManager);
+    mgr->AddFileManager(fileManager.clonePtr());
   }
 
   mFileManager = std::move(fileManager);
@@ -22260,7 +22267,7 @@
   mDatabase = MakeRefPtr<Database>(
       static_cast<Factory*>(Manager()), mCommonParams.principalInfo(),
       mOptionalContentParentId, mGroup, mOrigin, mTelemetryId,
-      mMetadata.clonePtr(), mFileManager, std::move(mDirectoryLock),
+      mMetadata.clonePtr(), mFileManager.clonePtr(), std::move(mDirectoryLock),
       mFileHandleDisabled, mChromeWriteAccessAllowed);
 
   if (info) {
@@ -23618,9 +23625,9 @@
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
   }
 
-  FileManager* fileManager = mDatabase->GetFileManager();
-
-  mFileInfo.init(fileManager->CreateFileInfo());
+  FileManager& fileManager = mDatabase->GetFileManager();
+
+  mFileInfo.init(fileManager.CreateFileInfo());
   if (NS_WARN_IF(!*mFileInfo)) {
     IDB_REPORT_INTERNAL_ERR();
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
@@ -23628,13 +23635,13 @@
 
   const int64_t fileId = (*mFileInfo)->Id();
 
-  const auto journalDirectory = fileManager->EnsureJournalDirectory();
+  const auto journalDirectory = fileManager.EnsureJournalDirectory();
   if (NS_WARN_IF(!journalDirectory)) {
     IDB_REPORT_INTERNAL_ERR();
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
   }
 
-  const auto journalFile = fileManager->GetFileForId(journalDirectory, fileId);
+  const auto journalFile = fileManager.GetFileForId(journalDirectory, fileId);
   if (NS_WARN_IF(!journalFile)) {
     IDB_REPORT_INTERNAL_ERR();
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
@@ -23645,13 +23652,13 @@
     return rv;
   }
 
-  const auto fileDirectory = fileManager->GetDirectory();
+  const auto fileDirectory = fileManager.GetDirectory();
   if (NS_WARN_IF(!fileDirectory)) {
     IDB_REPORT_INTERNAL_ERR();
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
   }
 
-  const auto file = fileManager->GetFileForId(fileDirectory, fileId);
+  const auto file = fileManager.GetFileForId(fileDirectory, fileId);
   if (NS_WARN_IF(!file)) {
     IDB_REPORT_INTERNAL_ERR();
     return NS_ERROR_DOM_INDEXEDDB_UNKNOWN_ERR;
@@ -24094,7 +24101,7 @@
                              const IndexMetadata& aMetadata)
     : VersionChangeTransactionOp(std::move(aTransaction)),
       mMetadata(aMetadata),
-      mFileManager(Transaction().GetDatabase()->GetFileManager()),
+      mFileManager(Transaction().GetDatabase()->GetFileManagerPtr()),
       mDatabaseId(Transaction().DatabaseId()),
       mObjectStoreId(aObjectStoreId) {
   MOZ_ASSERT(aObjectStoreId);
@@ -25343,7 +25350,7 @@
 
   if (mDataOverThreshold) {
     mStoredFileInfos.EmplaceBack(StoredFileInfo::CreateForStructuredClone(
-        aTransaction->GetDatabase()->GetFileManager()->CreateFileInfo(),
+        aTransaction->GetDatabase()->GetFileManager().CreateFileInfo(),
         MakeRefPtr<SCInputStream>(mParams.cloneInfo().data().data)));
   }
 
@@ -25542,12 +25549,9 @@
 
       const auto inputStream = inputStreamOrErr.unwrap();
 
-      auto* const fileManager = Transaction().GetDatabase()->GetFileManager();
-      MOZ_ASSERT(fileManager);
-
       if (inputStream) {
         if (fileHelper.isNothing()) {
-          fileHelper.emplace(RefPtr{fileManager});
+          fileHelper.emplace(Transaction().GetDatabase()->GetFileManagerPtr());
           rv = fileHelper->Init();
           if (NS_WARN_IF(NS_FAILED(rv))) {
             IDB_REPORT_INTERNAL_ERR();
@@ -25580,7 +25584,9 @@
         }
         if (NS_WARN_IF(NS_FAILED(rv))) {
           // Try to remove the file if the copy failed.
-          nsresult rv2 = fileManager->SyncDeleteFile(*file, *journalFile);
+          nsresult rv2 =
+              Transaction().GetDatabase()->GetFileManager().SyncDeleteFile(
+                  *file, *journalFile);
           if (NS_WARN_IF(NS_FAILED(rv2))) {
             return rv;
           }
@@ -25834,7 +25840,7 @@
   bool hasResult;
   while (NS_SUCCEEDED((rv = stmt->ExecuteStep(&hasResult))) && hasResult) {
     auto cloneInfoOrErr = GetStructuredCloneReadInfoFromStatement(
-        &*stmt, 1, 0, *mDatabase->GetFileManager());
+        &*stmt, 1, 0, mDatabase->GetFileManager());
     if (NS_WARN_IF(cloneInfoOrErr.isErr())) {
       return Err(cloneInfoOrErr.unwrapErr());
     }
@@ -26419,7 +26425,7 @@
   bool hasResult;
   while (NS_SUCCEEDED((rv = stmt->ExecuteStep(&hasResult))) && hasResult) {
     auto cloneInfoOrErr = GetStructuredCloneReadInfoFromStatement(
-        &*stmt, 1, 0, *mDatabase->GetFileManager());
+        &*stmt, 1, 0, mDatabase->GetFileManager());
     if (NS_WARN_IF(cloneInfoOrErr.isErr())) {
       return cloneInfoOrErr.unwrapErr();
     }